proposals/0011-renditions-list.md

```diff
@@ -0,0 +1,172 @@
+
+```js
+// assuming a video is already loaded
+const video = document.querySelector('video');
+
+const videoTrack = video.videoTracks[0];
+const videoRenditions = videoTrack.renditions;
+
+// only keep renditions enabled that are of height less than or requal to 720
+Array.from(videoRenditions).forEach(rendition => {
+    rendition.enabled = rendition.height <= 720;
+});
+```
+
+To change the selected rendition, you can set the `selectedIndex` on the renditions list.
+```js
+// assuming a video is already loaded
+const video = document.querySelector('video');
+
+const videoTrack = video.videoTracks[0];
+const videoRenditions = videoTrack.renditions;
+
+// switch to the 3rd rendition
+videoRenditions.selectedIndex = 2;
+```
+
+Changing the `selectedIndex` manually will trigger a `change` event on the renditions list. The ABR algorithm changing the selected rendition should also trigger a `change` event on the list object.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Below is WebIDL definitions for the changes and some explanation around it.
+
+```ts
+partial interface VideoTrack {
+  readonly attribute VideoRenditionList renditions;
+};
+```
+
+`VideoRenditionList` is similar to [`VideoTrackList`](https://html.spec.whatwg.org/multipage/media.html#videotracklist).
+The main difference aside from naming is the addition of a read-write `selectedIndex` property.
+This value will default to `-1` when no rendition is selected, such as before metadata has loaded.
+Otherwise, it will reflect the index of the currently selected rendition.
+Setting the value to a number will request a rendition change to that rendition.
+
+When the rendition changes, either via manually setting the `selectedIndex`, or automatically via the ABR algorithm, a `change` event should be triggered.
+
+```ts
+interface VideoRenditionList : EventTarget {
+  readonly attribute unsigned long length;
+  getter VideoRendition (unsigned long index);
+  VideoRendition? getRenditionById(DOMString id);
+  attribute long selectedIndex;
+
+  attribute EventHandler onchange;
+  attribute EventHandler onaddrendition;
+  attribute EventHandler onremoverendition;
+};
+```
+
+Then each `VideoRendition` is includes some metadata on the rendition, like width, height, bitrate, and codec.
+It also includes a boolean to determine whether this rendition is enabled or not.
+This allowed the ABR algorithms to know whether they should be able to switch to this rendition or not.
+
+If the currently active rendition is disabled, it should request a rendition change.
+
+```ts
+interface VideoRendition {
+  readonly attribute DOMString id;
+  readonly attribute unsigned long width;
+  readonly attribute unsigned long height;
+  readonly attribute unsigned long bitrate;
+  readonly attribute unsigned long frameRate;
+  readonly attribute unsigned long codec;
```

Comment:
should this be only the video codec? How to handle differing audio tracks for otherwise similar renditions?

Reply:
also, this should be a DOMString or something more relevant rather than a long.

Reply:
The easier path forward would be to simply omit any audio codec. The more robust would be to identify whether or not the audio track (in the container sense) is in-mux with the video track (in the container sense). If yes, we list both. If no, (e.g. with AUDIO groups for HLS or separate audio AdaptationSets for MPEG-DASH) we don't include the audio codec. I'm cool with either approach here.

Related: Maybe this should be the mime codec instead? That would make it consistent with: https://developer.mozilla.org/en-US/docs/Web/API/MediaSource/addSourceBuffer#parameters & https://developer.mozilla.org/en-US/docs/Web/HTML/Element/source#attributes

Reply:
yeah, maybe it should be `type` and match the source element type attribute and the `addSourceBuffer` param. Consistency is a plus, though, it does make it a bit more annoying since then users would need to parse out the codecs from the type.